### PR TITLE
sriov, presubmits, kubevirt, Update bootstrap image

### PR DESCRIPTION
**Pull request by oshoval**
As part of the effort to migrate project-infra images
from `docker.io` to `quay.io`,
update the bootstrap image of kubevirt sriov presubmit job
to use a newer one, which is hosted on `quay.io`.

Use a quay.io/kubevirtci/golang
in order to align kubevirt job to kubevirtci job,
and to have golang easier.
    
Since the new image doesn't have apt, but already has
`gettext-base`, update the job run command accordingly.
Remove `KUBEVIRT_PROVIDER` because `TARGET` is enough,
`KUBEVIRT_PROVIDER` will be set accordingly
by automation/test.s according to `TARGET`.

Add `set -e` in order to fail fast in case the job command
has an error, and not mask those errors.

Signed-off-by: Or Shoval <oshoval@redhat.com>

### CODE REVIEW

**Comment by oshoval** [comment]
https://prow.apps.ovirt.org/view/gcs/kubevirt-prow/pr-logs/pull/kubevirt_project-infra/1065/rehearsal-pull-kubevirt-e2e-kind-1.17-sriov/1373545235023400960

Passed

**Review by EdDev — COMMENTED** [review]
Please update the commit message with the reasoning and something that expresses what was updated.

**Comment by oshoval** [comment]
Addressed comments

**Review by EdDev — APPROVED** [review]

**Review by ormergi — CHANGES_REQUESTED** [review]
Please follow this PR example https://github.com/kubevirt/project-infra/pull/1050
and change the `command` and `env` section.
These jobs should be almost identical

**Comment by oshoval** [comment]
> Please follow this PR example #1050
> and change the `command` and `env` section.
> These jobs should be almost identical

This PR is for changing the bootstrap, nothing more.

Its not a go image, so i am not sure its the same state where you could use `GO_GIMMIE`
moreover, i did it already for the env vars that have a fixed value.
The env var that was left is the PATH, since its doesn't have a fixed value, it might look weird with the setting on PATH that depends on PATH.

I can try those two changes, but at least the `GO_GIMMIE` might not work.

**Comment by oshoval** [comment]
It fails
https://prow.apps.ovirt.org/view/gcs/kubevirt-prow/pr-logs/pull/kubevirt_project-infra/1065/rehearsal-pull-kubevirt-e2e-kind-1.17-sriov/1373586003956076544
`/bin/bash: go: command not found`
It's not a go image as the one in the example you refere, hence i can't do that.
In case there is something else, please be specific.

Reverting

**Comment by oshoval** [comment]
The failure is for the run with GO_GIMMIE
`11:03:21: /home/prow/go/src/github.com/kubevirt/kubevirt/cluster-up//cluster/kind-k8s-sriov-1.17.0/config_sriov.sh: line 192: go: command not found`
https://prow.apps.ovirt.org/view/gcs/kubevirt-prow/pr-logs/pull/kubevirt_project-infra/1065/rehearsal-pull-kubevirt-e2e-kind-1.17-sriov/1373586003956076544

**Comment by oshoval** [comment]
/rehearse

(reverted `GO_GIMMIE`, it passed before when it was without it, lets run again)

**Comment by ormergi** [comment]
You need to use quay.io/kubevirtci/golang:v20210316-d295087 for the go-gimme to work
This is image that is used on the check-cluster-up lane

**Comment by oshoval** [comment]
> You need to use quay.io/kubevirtci/golang:v20210316-d295087 for the go-gimme to work
> This is image that is used on the check-cluster-up lane

I know, and this is what i am saying, but i am not sure this is the image we want here, 
since almost all kubevirt uses a bootstrap image, i am not sure we want sriov to be different.

**Comment by oshoval** [comment]
I think we shouldn't block about the go,
its not in the direct scope of this PR, we can open other PR if needed for that.
This PR intends to align the boostrap image to be on quay.io, and i find it better to have it as most of the other images.
Its true that it saves time, but we might consider on the other hand to install go in the bootstrap so all the jobs that need go will have the option ? (on other PR as well)

I would be glad if we can divide it to parts and focus on the current topic please.

**Comment by oshoval** [comment]
@qinqon @fgimenez @ormergi 
Please suggest, is it better that we will use `quay.io/kubevirtci/bootstrap:v20210311-09ebaa2` on kubevirt, as most of the jobs do,
and install go ourself.
or use the `quay.io/kubevirtci/golang:v20210316-d295087` bootstrap image as kubevirtci sriov job does and then we can use `GO_GIMMIE`.

2 Points here:
1. Its indeed better that sriov kubevirtci and kubevirt will be the same, in one way or another.
2. Worth considering (if possible), that the `quay.io/kubevirtci/bootstrap:v20210311-09ebaa2` will have go installed, so every job that need go will be able to use `GO_GIMMIE`.

Thanks

**Comment by oshoval** [comment]
Rebase
since point [1] here is good enough reason, i will update the image
https://github.com/kubevirt/project-infra/pull/1065#issuecomment-803834963

**Comment by oshoval** [comment]
Updated to use `quay.io/kubevirtci/golang:v20210316-d295087`

File: github/ci/prow-deploy/files/jobs/kubevirt/kubevirt/kubevirt-presubmits.yaml
Comment (ormergi) [inline github/ci/prow-deploy/files/jobs/kubevirt/kubevirt/kubevirt-presubmits.yaml:341]: nit: Add && between the trap and automation/test.sh
`trap "echo teardown && make cluster-down" EXIT ERR SIGINT SIGTERM && automation/test.sh`

Comment (oshoval) [reply github/ci/prow-deploy/files/jobs/kubevirt/kubevirt/kubevirt-presubmits.yaml:341]: trap can't fail, so i don't see a reason to.
its also not scalable (to add commands) this way, its better to be a command list, than a '&&' list of one command.

Comment (ormergi) [reply github/ci/prow-deploy/files/jobs/kubevirt/kubevirt/kubevirt-presubmits.yaml:341]: The problem is that If something goes wrong the job will continue silently.
For example, this is exactly what happen with stiov check cluster-up lane last week
I think you already familiar with #1050, we could prevent the false-positive report if there was && between the trap to other commands.
Not scalable? it dosent prevent one from adding more commands.

Comment (oshoval) [reply github/ci/prow-deploy/files/jobs/kubevirt/kubevirt/kubevirt-presubmits.yaml:341]: Trap can't fail.
Its a bad practice to '&&' it to other commands, i am talking about it specifically.
one can add commands below it and '&&' them with themself if needed.

Yes i am familiar with it, and found it also, before knowing about that PR.

We are creating here a bad practice just because with other commands there is a possible to fail.
In case one adds a command, he will need to think if to '&&' it, instead just doing it, just because its there already.

Comment (oshoval) [reply github/ci/prow-deploy/files/jobs/kubevirt/kubevirt/kubevirt-presubmits.yaml:341]: What about adding set -e ?
automation/test.sh already have set -e is it won't change the behavior.

```
            set -e;
            trap "echo teardown && make cluster-down" EXIT ERR SIGINT SIGTERM;
            automation/test.sh;
```

/cc @qinqon since you suggested this '&&' first
please advice, and lets wrap it

Comment (qinqon) [reply github/ci/prow-deploy/files/jobs/kubevirt/kubevirt/kubevirt-presubmits.yaml:341]: I am ok with `set -e`

**Comment by oshoval** [comment]
Added `set -e`

**Comment by oshoval** [comment]
Rebase

File: github/ci/prow-deploy/files/jobs/kubevirt/kubevirt/kubevirt-presubmits.yaml
Comment (ormergi) [reply github/ci/prow-deploy/files/jobs/kubevirt/kubevirt/kubevirt-presubmits.yaml:378]: Please dont add set -e, we already have it on the test.sh script and we put a trap on the command section
that another thing to worry about.

> Trap can't fail.
Its a bad practice to '&&' it to other commands, i am talking about trap specifically.
one can add commands below it and '&&' them with themself if needed.

In case of another bootstrap image change, the shell could also change and bring unexpected behavior.
e.g: trap is not supported, trap syntax change..

Comment (oshoval) [reply github/ci/prow-deploy/files/jobs/kubevirt/kubevirt/kubevirt-presubmits.yaml:378]: Sorry, i don't see how trap is going to stop be supported,
and the set -e here is for the command of the job
i find it much better than using && especially when the trap can't fail.
One don't need to worry on this about, on the other hand, if anybody would need to worry about adding && its the not thing to have, this set -e solve it.

**Comment by oshoval** [comment]
https://prow.apps.ovirt.org/view/gcs/kubevirt-prow/pr-logs/pull/kubevirt_project-infra/1065/rehearsal-pull-kubevirt-e2e-kind-1.17-sriov/1373935305215709184
Passed

File: github/ci/prow-deploy/files/jobs/kubevirt/kubevirt/kubevirt-presubmits.yaml
Comment (ormergi) [reply github/ci/prow-deploy/files/jobs/kubevirt/kubevirt/kubevirt-presubmits.yaml:378]: We have `trap .. ERR` to catch errors, in case more commands should be added to this job it should be added after the trap line. 
Adding set -e is redundant

Comment (oshoval) [reply github/ci/prow-deploy/files/jobs/kubevirt/kubevirt/kubevirt-presubmits.yaml:378]: `set -e` cause that a command with an error will exit the "script"
trap ... ERR on its own won't help
hence its not redundant
run this and see done is printed, add set -e and see it doesn't

```
#!/bin/bash

trap "echo teardown" ERR;

false

echo done
```

Comment (qinqon) [reply github/ci/prow-deploy/files/jobs/kubevirt/kubevirt/kubevirt-presubmits.yaml:378]: As far as I know "-e" modifier and trap are different things, one make the script fail if one of the commands fail and the other i s use to "arm" the error signal so they complement each other.

I think we can add `set -e` and `trap` so we don't care if the commands are done with `&&` or  `;`

**Comment by oshoval** [comment]
The error of rehearsal-pull-kubevirt-e2e-k8s-1.18
is unrelated, i stopped it by the rebase,
from my side, all is set to proceed please.

**Comment by oshoval** [comment]
/cc @fgimenez 
can you take a look please?

File: github/ci/prow-deploy/files/jobs/kubevirt/kubevirt/kubevirt-presubmits.yaml
Comment (ormergi) [reply github/ci/prow-deploy/files/jobs/kubevirt/kubevirt/kubevirt-presubmits.yaml:378]: > This way we can add lines without infinite '&&'.
Its a bad practice to '&&' commands, especially the trap, which can't fail.
The code should be nice and easy to maintain.

This is not the example I suggested, on your examples replace the ; with && on the trap with and you will see the script will exit.

> As far as I know "-e" modifier and trap are different things, one make the script fail if one of the commands fail and the other i s use to "arm" the error signal so they complement each other.

I think we should test it first

Deleting KIND infra cluster w/o using `kind` binary may cause resources leak (cluster-down use it)
thus its important to test SRIOV lane jobs before merging.
After we finish with this job we can continue with release branches jobs.
Please inject an error so we could see that it does fail and cluster-down is invoked once.

**Comment by ormergi** [comment]
Deleting KIND infra cluster w/o using `kind` binary may cause resources leak 
cluster-down use kind binary to delete the cluster.
Since you change how errors are handheld its important to test it before merging.
Could you please inject an error so we could see that it does fail and cluster-down is invoked **once**.

After we finish with this job we can continue with release branches jobs.

File: github/ci/prow-deploy/files/jobs/kubevirt/kubevirt/kubevirt-presubmits.yaml
Comment (oshoval) [reply github/ci/prow-deploy/files/jobs/kubevirt/kubevirt/kubevirt-presubmits.yaml:378]: I tested it here
https://github.com/kubevirt/project-infra/pull/1068
it runs the trap.

**Comment by oshoval** [comment]
> Deleting KIND infra cluster w/o using `kind` binary may cause resources leak
> cluster-down use kind binary to delete the cluster.
> Since you change how errors are handheld its important to test it before merging.
> Could you please inject an error so we could see that it does fail and cluster-down is invoked **once**.
> 
> After we finish with this job we can continue with release branches jobs.

This PR should not touch the release branches, you are welcome to do so on other PR.
I tested it, see and lets please continue at https://github.com/kubevirt/project-infra/pull/1065#discussion_r598657887
if needed

**Comment by ormergi** [comment]
> This PR should not touch the release branches, you are welcome to do so on other PR.
> Release branches should not be touched unless mandatory.
> 

I didnt asked you to do so.

> I tested it, see and lets please continue at [#1065 (comment)](https://github.com/kubevirt/project-infra/pull/1065#discussion_r598657887)
> if needed
> 
> It could run twice, or run without kind binary even before, if needed it can be fixed on a follow PR, it doesnt related to this PR (nor part of project-infra).

What is the point in merging somthing that we would need to fix later?
We dont want cluster-down to run twice it should run once and it must use kind binary.

**Comment by oshoval** [comment]
> > This PR should not touch the release branches, you are welcome to do so on other PR.
> > Release branches should not be touched unless mandatory.
> 
> I didnt asked you to do so.
> 
> > I tested it, see and lets please continue at [#1065 (comment)](https://github.com/kubevirt/project-infra/pull/1065#discussion_r598657887)
> > if needed
> > It could run twice, or run without kind binary even before, if needed it can be fixed on a follow PR, it doesnt related to this PR (nor part of project-infra).
> 
> What is the point in merging something that we would need to fix later?
> We dont want cluster-down to run twice it should run once and it must use kind binary.

This PR doesnt present a regression, it could run twice the cluster-down as well as before,
this PR is intended to update the bootstrap image thats all.

**Comment by oshoval** [comment]
Rebased

There is no valid reason so far to block this PR.
There is no regression created by this PR.
We had always potential double cluster-down, (one in the job, and one in automation/test.sh)
i can remove the one in the job if needed.

**Comment by ormergi** [comment]
I ran few tests and there are two options using && between commands or adding `set -e`
I guess it matter of taste.
  
> I even not sure there is a problem with running it twice, or that it can run it without the binary,
> about running it twice - it wont find it in the 2nd time.
> about running it without the binary - it means it even didnt create it.

I verified it while back and it is recommended to use kind binary for teardown (to prevent resources leak due to leftovers, also we use kind seamlessly by cluster down)
It will run cluster-down twice if there is an error because we also specify EXIT flag (but this is out of this PR scope to change it)
I guess running cluster-down wont harm much.

/lgtm

**Comment by oshoval** [comment]
Cool

@fgimenez 
can you take a look please?

File: github/ci/prow-deploy/files/jobs/kubevirt/kubevirt/kubevirt-presubmits.yaml
Comment (ormergi) [inline github/ci/prow-deploy/files/jobs/kubevirt/kubevirt/kubevirt-presubmits.yaml:376]: Since you mentioned that we need add `;` at the end of each command
if we would use `|-` instead, we wont need to add `;`
(Its just a suggestion I understand if you dont want to do it on this PR)

**Comment by oshoval** [comment]
Rebased

File: github/ci/prow-deploy/files/jobs/kubevirt/kubevirt/kubevirt-presubmits.yaml
Comment (oshoval) [reply github/ci/prow-deploy/files/jobs/kubevirt/kubevirt/kubevirt-presubmits.yaml]: Done, used '|' because we we don't care for 'strip'
its not that we have additional empty lines.

**Comment by oshoval** [comment]
Updated to use '|' instead of '>' 
so we don't need ';' between new lines

File: github/ci/prow-deploy/files/jobs/kubevirt/kubevirt/kubevirt-presubmits.yaml
Comment (oshoval) [inline github/ci/prow-deploy/files/jobs/kubevirt/kubevirt/kubevirt-presubmits.yaml:378]: since i removed `export KUBEVIRT_PROVIDER=kind-k8s-sriov-1.17.0`
the trap here tries to run the wrong cluster-down
we have already a cluster-down in automation/test.sh
so i will remove this one ok ?
alternatively i can readd the `export KUBEVIRT_PROVIDER=kind-k8s-sriov-1.17.0`

**Comment by oshoval** [comment]
first push - rebase
second - add KUBEVIRT_PROVIDER so the cluster-down in the command will run the right target

**Review by fgimenez — APPROVED** [review]
/lgtm
/approve

**Pull request merged**
